src/ClientMessage.ts

```diff
@@ -15,248 +15,231 @@
     }
 
-    readUInt8(): number {
-        const value = this.buffer.readUInt8(this.cursor);
-        this.cursor += BitsUtil.BYTE_SIZE_IN_BYTES;
-        return value;
+    getPartitionId(): number {
+        return this.startFrame.content.readInt32LE(PARTITION_ID_OFFSET);
     }
 
-    readInt32(): number {
-        const value = this.buffer.readInt32LE(this.cursor);
-        this.cursor += BitsUtil.INT_SIZE_IN_BYTES;
-        return value;
+    setPartitionId(partitionId: number): void {
+        this.startFrame.content.writeInt32LE(partitionId, PARTITION_ID_OFFSET);
     }
 
-    readLong(): Long {
-        const value = this.readLongInternal(this.cursor);
-        this.cursor += BitsUtil.LONG_SIZE_IN_BYTES;
-        return value;
+    getHeaderFlags(): number {
+        return this.startFrame.flags;
     }
 
-    readString(): string {
-        const length = this.buffer.readInt32LE(this.cursor);
-        this.cursor += BitsUtil.INT_SIZE_IN_BYTES;
-        const value = this.buffer.toString('utf8', this.cursor, this.cursor + length);
-        this.cursor += length;
-        return value;
+    isRetryable(): boolean {
+        return this.retryable;
     }
 
-    readBuffer(): Buffer {
-        const size = this.buffer.readUInt32LE(this.cursor);
-        this.cursor += BitsUtil.INT_SIZE_IN_BYTES;
-        const result = this.buffer.slice(this.cursor, this.cursor + size);
-        this.cursor += size;
-        return result;
+    setRetryable(retryable: boolean): void {
+        this.retryable = retryable;
     }
 
-    isComplete(): boolean {
-        return (this.cursor >= BitsUtil.HEADER_SIZE) && (this.cursor === this.getFrameLength());
+    getConnection(): ClientConnection {
+        return this.connection;
     }
 
-    readMapEntry(): any {
-        // TODO
+    setConnection(connection: ClientConnection): void {
+        this.connection = connection;
     }
 
-    private writeLongInternal(value: any, offset: number): void {
-        if (!Long.isLong(value)) {
-            value = Long.fromValue(value);
+    getTotalFrameLength(): number {
+        let frameLength = 0;
+        let currentFrame = this.startFrame;
+        while (currentFrame != null) {
+            frameLength += currentFrame.getLength();
+            currentFrame = currentFrame.next;
         }
+        return frameLength;
+    }
 
-        this.buffer.writeInt32LE(value.low, offset);
-        this.buffer.writeInt32LE(value.high, offset + 4);
+    merge(fragment: ClientMessage): void {
+        // Ignore the first frame of the fragment since first frame marks the fragment
+        this.endFrame.next = fragment.startFrame.next;
+        this.endFrame = fragment.endFrame;
     }
 
-    private readLongInternal(offset: number): Long {
-        const low = this.buffer.readInt32LE(offset);
-        const high = this.buffer.readInt32LE(offset + 4);
-        return new Long(low, high);
+    copyWithNewCorrelationId(correlationId: any): ClientMessage {
```

Comment:
nit: in all calls made to this method we set `correlationId` argument to `-1` (see `src/ListenerService.ts`). So, probably we can remove this argument and always set correlation id of the copied message to `-1` implicitly. WDYT?